### PR TITLE
Flow inference values across generic method return types.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Say you have:

```c#
    private void M()
    {
        bool merged = X(Test());   //<-- trying to infer a return type for 'Test'
    }

    private T X<T>(T t) { return t; }
```

We'll now flow types appropriate so that we:
1. Use the variable declaration type (bool) to flow into the initialize expression.
2. See that the method is generic. 
3. Attempt to instantiate the method such that its return type will be 'bool'. In this case this will cause us to realize that 'T' should be instantiated to bool.  

As such, we'll realize that 'Test' needs to be 'bool' returning to make the above code correct.

This also works across complex scenarios like:

```c#
    bool result = await Task.Run(() => Test());   //<-- determining return type of 'Test'
```

Here we will appropriate flow 'bool' all the way into the lambda and we'll determine it is the appropriate return type for 'Test'.

Fixes: https://github.com/dotnet/roslyn/issues/4233

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Reviewers: @balajikris, @basoundr, @brettfo, @davkean, @DustinCampbell, @dpoeschl, @Pilchie

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 1]
```diff
@@ -605,9 +605,129 @@ private IEnumerable<ITypeSymbol> InferTypeInAttributeArgument(int index, IEnumer
...
 
+            private IMethodSymbol Instantiate(IMethodSymbol method, IList<ITypeSymbol> invocationTypes)
+            {
+                // No need to instantiate if this isn't a generic method.
+                if (method.TypeArguments.Length == 0)
+                {
+                    return method;
+                }
+
+                // Can't infer the type parameters if this method doesn't have a return type.
```
Comment (dpoeschl) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: As we discussed offline, we infer the type parameters from other parameters elsewhere in the type inferrer, so this comment is true within the context of what we're specifically trying to do here, which is flow from the return type to the parameters. Can you add that clarification to this comment, or more preferably to the comments in InferTypeInArgument above?

**Comment by dpoeschl** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: Sure!

**Pull request merged**